### PR TITLE
Questionnaire table pagination

### PR DESCRIPTION
**Pull request by tgandrews**
### What is the context of this PR?
Add pagination to the questionnaires table

### How to review 
1. Ensure you can paginate through the questionnaires
2. Ensure that the page number is maintained across loads (in app and page refreshes)

### CODE REVIEW

**Comment by tgandrews** [comment]
[Pagination (6)](https://trello.com/c/yuRVsMZU/1072-pagination-6)

File: eq-author/src/App/QuestionnairesPage/QuestionnairesTable/index.js [hunk 1]
```diff
@@ -38,120 +43,163 @@ const Panel = styled.div`
...
+                questionnaire={questionnaire}
+                onDeleteQuestionnaire={handleDeleteQuestionnaire}
+                onDuplicateQuestionnaire={onDuplicateQuestionnaire}
+                isLastOnPage={state.currentPage.length === index + 1}
+              />
+            );
+          })}
+        </TransitionGroup>
+      </Table>
+      <TableFooter>
```
Comment (heofs) [inline eq-author/src/App/QuestionnairesPage/QuestionnairesTable/index.js]: I noticed on [the prototype](https://prototype-sharing--author-app.netlify.com/#/), the footer with the pagination and results text is displayed outside the table. Should this be moved to outside the panel?

Comment (tgandrews) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesTable/index.js]: Yeah, as discussed requires a minor refactor to get this working. But you are correct and I'm making the change now

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/index.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import styled from "styled-components";
+import PropTypes from "prop-types";
+
+import { colors } from "constants/theme";
+
+import PaginationNav from "./PaginationNav";
+
+const TableFooter = styled.div`
+  display: flex;
+  padding: 1em;
```
Comment (heofs) [inline eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/index.js]: The results text should be aligned with the table. You could drop the padding and use `margin: 1em 0;` to get the layout the same as the prototype.

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/QuestionnairesTable/Row/index.test.js [hunk 3]
```diff
@@ -137,15 +139,41 @@ describe("Row", () => {
...
-      props.questionnaire
-    );
-    expect(stopPropagation).toHaveBeenCalled();
+      expect(wrapper.find(DeleteConfirmDialog).props()).toMatchObject({
+        isOpen: true,
+      });
+      expect(stopPropagation).toHaveBeenCalled();
+    });
+
+    it("call onDelete when the dialog is confirmed", () => {
```
Comment (heofs) [inline eq-author/src/App/QuestionnairesPage/QuestionnairesView/QuestionnairesTable/Row/index.test.js]: should it be "should call..."?

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+const Buttons = styled.nav`
+  margin: 0 auto;
+  display: flex;
+  justify-content: center;
+  text-align: center;
+`;
+
+const Pagination = ({ currentPageIndex, pageCount, onPageChange }) => {
+  return (
+    <Buttons role="navigation" aria-label="Questionnaires navigation">
```
Comment (hamishtaplin) [inline eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js]: You don't need a `role=navigation` for a `nav` element as it has those semantics already. Perhaps rename `Button` to `Nav` to make this more obvious?

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/index.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    STORAGE_KEY,
+    STORED_KEYS,
+    reducer,
+    { currentPageIndex: 0 },
+    buildInitialState(questionnairesRef.current)
+  );
+
+  if (questionnaires !== questionnairesRef.current) {
+    dispatch({
+      type: "SET_QUESTIONNAIRES",
```
Comment (hamishtaplin) [inline eq-author/src/App/QuestionnairesPage/QuestionnairesView/index.js]: Better to declare these action types as constants? I would export from the reducer, personally, but they could live in their own file somewhere.

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/index.test.js [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+
+        expect(getByText("2 of 2")).toBeTruthy();
+
+        const previousPage = getByText("Go to previous page");
+        fireEvent.click(previousPage);
+
+        expect(getByText("1 of 2")).toBeTruthy();
+      });
+
+      it("should be impossible to a page before the first one", () => {
```
Comment (heofs) [inline eq-author/src/App/QuestionnairesPage/QuestionnairesView/index.test.js]: This and line 330. Did you mean "should be impossible to navigate"?

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/reducer.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      questionnaires.find(q => q.id === deletedQuestionnaire.id)
+    ) + 1;
+
+  // If the last one is being removed then focus the one before that
+  const nextIndex =
+    possibleNextIndex > questionnaires.length - 1
+      ? questionnaires.length - 2
+      : possibleNextIndex;
+
+  // We have to account to set focusedId to undefined when there are no
```
Comment (heofs) [inline eq-author/src/App/QuestionnairesPage/QuestionnairesView/reducer.js]: Remove "account to"?

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js (hunk 4 above)
Comment (tgandrews) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js]: According to this article they are suggest adding it to make it clearer. https://www.a11ymatters.com/pattern/pagination/ .

I did think it was weird. Thoughts?

The name change sounds reasonable either way.

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/index.js (hunk 5 above)
Comment (tgandrews) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesView/index.js]: Yeah, I'll move them to the reducer

File: eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js (hunk 4 above)
Comment (tgandrews) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js]: Have removed it as other articles don't suggest using it. http://web-accessibility.carnegiemuseums.org/code/navigation/

Comment (tgandrews) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js]: Another article suggest adding it http://web-accessibility.carnegiemuseums.org/code/navigation/

Comment (tgandrews) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js]: MDN says it isn't necessary
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Navigation_Role

Comment (tgandrews) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js]: I've removed it. I trust MDN and semantically that seems the most logical. Until we have a screen reader which we know of that can't cope lets stick with the simpler version

Comment (hamishtaplin) [reply eq-author/src/App/QuestionnairesPage/QuestionnairesView/Footer/PaginationNav.js]: That was a wild ride

**Pull request merged**